### PR TITLE
[RLlib] BC RLModule

### LINKED ISSUE #37776: [RLlib] Enabling RLModule by default on BC
### Description

We need to enable the new stack by default. 

### Use case

_No response_

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

Implements BC with RLModule API. 

Results for both, the "old" `ModelV2` implementation (orange:torch, blue:tf2) and the "new" `RLModule` implementation (red:torch, lightblue:tf2) can be analyzed in the [TensorBoard](https://tensorboard.dev/experiment/EQHMzV1SQeW6pwpFlJ0AXw/)

## Related issue number

Closes #37776 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/algorithms/bc/bc_rl_module.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    @override(RLModule)
+    def get_initial_state(self) -> Any:
+        if hasattr(self.encoder, "get_initial_state"):
+            return self.encoder.get_initial_state()
+        else:
+            return {}
+
+    @override(RLModule)
+    def output_specs_inference(self) -> SpecType:
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/bc_rl_module.py:44]: @kouroshHakha For the specs of all of these RL Modules, I think they should just return a `Spec`, right?

Right now, we have a mixture of `SpecType` and `SpecDict` which are both not correct.
This should be any Spec, just something that I can call `Spec.validate` on.

File: rllib/algorithms/bc/tests/test_bc_rl_module.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    @classmethod
+    def tearDownClass(cls):
+        ray.shutdown()
+
+    def test_bc_compilation_and_learning_from_offline_file(self):
+        """Test whether BC can be built with all frameworks.
+
+        And learns from a historic-data file (while being evaluated on an
+        actual env using  evaluation_num_workers > 0).
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/tests/test_bc_rl_module.py:30]:
```suggestion
        actual env using evaluation_num_workers > 0).
```

File: rllib/algorithms/bc/tests/test_bc_rl_module.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    framework_iterator,
+)
+
+tf1, tf, tfv = try_import_tf()
+
+
+class TestBRLModuleC(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        ray.init(local_mode=True)
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/tests/test_bc_rl_module.py:20]: This needs to go :)

File: rllib/algorithms/bc/tests/test_bc_rl_module.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+from ray.rllib.utils.test_utils import (
+    check_compute_single_action,
+    check_train_results,
+    framework_iterator,
+)
+
+tf1, tf, tfv = try_import_tf()
+
+
+class TestBRLModuleC(unittest.TestCase):
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/tests/test_bc_rl_module.py:17]: Can we consolidate this test with the other one that does not use RL Modules?
They should be the same (save for the rl module setting), so there should be a loop that loops over `rl_modules=True/False`. By doing that we can be sure/demonstrate that the task is the exact same so that the new implementation is equivalent. We also get the chance to mitigate code divergence between the old and new testing code in the near future.

File: rllib/algorithms/marwil/marwil_tf_policy.py [hunk 5]
```diff
@@ -55,7 +55,7 @@ def postprocess_trajectory(
             # Create an input dict according to the Model's requirements.
             index = "last" if SampleBatch.NEXT_OBS in sample_batch else -1
             input_dict = sample_batch.get_single_step_input_dict(
-                self.model.view_requirements, index=index
+                self.view_requirements, index=index
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/marwil/marwil_tf_policy.py:58]: What's the reason behind this change? Can we get rid of it?

File: rllib/algorithms/bc/tf/bc_tf_learner.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    @override(TfLearner)
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        hps: LearnerHyperparameters,
+        batch: NestedDict,
+        fwd_out: Mapping[str, TensorType]
+    ) -> TensorType:
+        if self.module[module_id].is_stateful():
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/tf/bc_tf_learner.py:35]: Since we are adding code here for the recurrent case, can you also include this in the test?
We do the same thing in PPO so there's a chance you can copy some code!

File: rllib/algorithms/bc/tf/bc_tf_rl_module.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+from ray.rllib.algorithms.bc.bc_rl_module import BCRLModule
+from ray.rllib.core.models.base import ENCODER_OUT, STATE_OUT
+from ray.rllib.core.rl_module.rl_module import RLModule
+from ray.rllib.core.rl_module.tf.tf_rl_module import TfRLModule
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.nested_dict import NestedDict
+from ray.rllib.utils.framework import try_import_tf
+
+tf1, tf, _ = try_import_tf()
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/tf/bc_tf_rl_module.py:12]: We don't need any tf here!
Same goes for the torch version!

@simonsays1980 Please consolidate the methods of both RL Modules into one `BCBaseRLModule`.
This `BCBaseRLModule` should have the docstrings that you currently have in the torch version.
If possible, the two framework-specific RL Modules should just be:

```
class BCTfRLModule(TfRLModule, BCRLModule):
    framework: str = "tf2"
```

and

```
class BCTorchRLModule(TorchRLModule, BCRLModule):
    framework: str = "torch"
```

File: rllib/algorithms/bc/torch/bc_torch_rl_module.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+from ray.rllib.core.rl_module.torch.torch_rl_module import TorchRLModule
+from ray.rllib.policy.sample_batch import SampleBatch
+from ray.rllib.utils.annotations import override
+from ray.rllib.utils.framework import try_import_torch
+from ray.rllib.utils.nested_dict import NestedDict
+
+torch, nn = try_import_torch()
+
+
+class BCTorchRLModule(TorchRLModule, BCRLModule):
```
Comment (ArturNiederfahrenhorst) [inline rllib/algorithms/bc/torch/bc_torch_rl_module.py:15]: The comment for the BCTfRLModule also applies here, please consolidate 🙂

**Review by ArturNiederfahrenhorst — CHANGES_REQUESTED** [review]
Great work! Thanks a lot @simonsays1980 

I have a couple of changes to request.
When you are confident that this is going well, can you also provide proof that this learns by running the cartpole learning test `cartpole-bc.yaml` and provide screenshots of the tensorboard that holds at least two runs that show the learning behaviour "old vs new"?

File: rllib/algorithms/bc/bc.py [hunk 9]
```diff
@@ -73,3 +136,49 @@ class BC(MARWIL):
...
+        if not self.config["_enable_rl_module_api"]:
+            # Using ModelV2.
+            return super().training_step()
+        else:
+            # Implement logic using RLModule and Learner API.
+            with self._timers[SAMPLE_TIMER]:
+                # TODO (simon): Check, if multi-agent setting could be used here
+                # as well.
+                # Sampling from offline data.
+                train_batch = synchronous_parallel_sample(worker_set=self.workers)
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/bc/bc.py:151]: @ArturNiederfahrenhorst Thanks for the clarification. I was unsure, if BC was implemented before for MA

File: rllib/algorithms/bc/bc_rl_module.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    @override(RLModule)
+    def get_exploration_action_dist_cls(self) -> Type[Distribution]:
+        return self.action_dist_cls
+
+    @override(RLModule)
+    def get_inference_action_dist_cls(self) -> Type[Distribution]:
+        return self.action_dist_cls
+
+    @override(RLModule)
+    def get_initial_state(self) -> Any:
```
[the remark this answers is not in the document]
Comment (simonsays1980) [reply rllib/algorithms/bc/bc_rl_module.py:37]: I took this from the `rl_module.py`:

```python
@OverrideToImplementCustomLogic
def get_initial_state(self) -> Any:
    """Returns the initial state of the module.

    This can be used for recurrent models.
    """
    return {}
```
But when I look into the `core/models/tf/encoder.py` I find indeed:

```python
@override(Model)
def get_initial_state(self):
    return {
        "h": tf.zeros((self.config.num_layers, self.config.hidden_dim)),
    }
```
As we use here the default encoder I think it is better to follow your suggestion and define the return type to `Union[dict, List[TensorType]]`.

File: rllib/algorithms/bc/tests/test_bc_rl_module.py (hunk 3 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/tests/test_bc_rl_module.py:20]: My good old friend lol

File: rllib/algorithms/marwil/marwil_tf_policy.py (hunk 5 above)
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil_tf_policy.py:58]: I agree this looks strange at first. The reason behind it is: 

- We still need the policies for the RL Module API - in this case the BC policy depends on the MARWIL one
- This policy calls the view requirements of the model - but the `ModelV2`. In the RL Module case `model` is a `RLModule` and does not have the attribute `view_requirements`; this throws an exception.
- To overcome this I call the view requirements from the policy, which should be the same I guess - any exceptions you can think of?

This should become obsolete as soon as we are deprecating MARWIL to `rllib_contrib` as then the BC policy must be independent of the MARWIL one and we do not need any postprocessing as we do not need returns.

File: rllib/algorithms/bc/bc_rl_module.py (hunk 1 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/bc_rl_module.py:44]: Following the example [here](https://docs.ray.io/en/releases-2.6.3/rllib/rllib-rlmodule.html#writing-custom-single-agent-rl-modules): 
```python
class DiscreteBCTorchModule(TorchRLModule):
    ...

    @override(TorchRLModule)
    def input_specs_exploration(self) -> SpecType:
        # Enforce that input nested dict to exploration method has a key "obs"
        # and within that key, it has a key "global" and "local". There should
        # also be a key "action_mask"
        return [("obs", "global"), ("obs", "local"), "action_mask"]
```
and the deifnition of `format_to_canonical_form()`.

The output of `output_spec_inference()` can be of a type that does not inherit from `Spec` as I understand it and therefore the `SpecType` is used that also allows a list of string or tuple of strings. 

We could possible use instead of the "simpler" list of string here, a `SpecDict` as the `SampleBatch.ACTION_DIST_INPUTS` will be `TensorType`s.

Comment (kouroshHakha) [reply rllib/algorithms/bc/bc_rl_module.py:44]: If I remember correctly SpecType is a superset of all possible format that we support which includes Spec and SpecDicts and normal dicts, etc.

Comment (simonsays1980) [reply rllib/algorithms/bc/bc_rl_module.py:47]: @kouroshHakha Thanks for the clarification!

File: rllib/algorithms/marwil/marwil_tf_policy.py (hunk 5 above)
Comment (simonsays1980) [reply rllib/algorithms/marwil/marwil_tf_policy.py:58]: @ArturNiederfahrenhorst Shall we go with it?

File: rllib/algorithms/bc/torch/bc_torch_rl_module.py (hunk 8 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/torch/bc_torch_rl_module.py:5]: Done

File: rllib/algorithms/bc/bc_rl_module.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/bc/bc_rl_module.py:47]: Ok so we should go with SpecType in the general case.

**Comment by ArturNiederfahrenhorst** [comment]
Taking over this PR to fix the LSTM issue that blocks it

File: rllib/algorithms/marwil/marwil_tf_policy.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/algorithms/marwil/marwil_tf_policy.py:58]: Yes, all good. This should indeed not pose a problem.

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Approved based on green tests

**Comment by ArturNiederfahrenhorst** [comment]
I've created an issue that describes what is needed to add recurrence to BC at https://github.com/ray-project/ray/issues/39640

**Comment by ArturNiederfahrenhorst** [comment]
For now, let's go without LSTM. Adding support for offline recurrent networks needs a little bit of design decision making and will impact other algorithms, too.

**Review by ArturNiederfahrenhorst — APPROVED** [review]

File: rllib/algorithms/bc/bc_catalog.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
+# __sphinx_doc_begin__
+import gymnasium as gym
+
+from ray.rllib.algorithms.ppo.ppo_catalog import _check_if_diag_gaussian
+from ray.rllib.core.models.catalog import Catalog
+from ray.rllib.core.models.configs import FreeLogStdMLPHeadConfig, MLPHeadConfig
+
```
Comment (sven1977) [inline rllib/algorithms/bc/bc_catalog.py:7]: nit: remove this empty line

File: rllib/algorithms/bc/tf/bc_tf_rl_module.py [hunk 12]
```diff
@@ -0,0 +1,6 @@
+from ray.rllib.algorithms.bc.bc_rl_module import BCRLModule
+from ray.rllib.core.rl_module.tf.tf_rl_module import TfRLModule
+
+
+class BCTfRLModule(TfRLModule, BCRLModule):
+    framework: str = "tf2"
```
Comment (sven1977) [inline rllib/algorithms/bc/tf/bc_tf_rl_module.py:6]: Dumb question: Should we move the definition `framework: str = "tf2"` (same for torch) into TfRLModule instead? And then just do:
```
class BCTfRLModule(TfRLModule, BCRLModule):
    pass
```
here?

File: rllib/algorithms/marwil/marwil_tf_policy.py (hunk 5 above)
Comment (sven1977) [reply rllib/algorithms/marwil/marwil_tf_policy.py:58]: fine with me, too

File: rllib/algorithms/bc/torch/bc_torch_learner.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        hps: LearnerHyperparameters,
+        batch: NestedDict,
+        fwd_out: Mapping[str, TensorType]
+    ) -> TensorType:
+        if self.module[module_id].is_stateful():
+            # In the RNN case, we expect incoming tensors to be padded to the maximum
```
Comment (sven1977) [inline rllib/algorithms/bc/torch/bc_torch_learner.py:34]: not: Can we move these comment blocks always on top of the `if` line? Same for the `else` line/comment. Thx :)

File: rllib/algorithms/bc/torch/bc_torch_learner.py [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    @override(TorchLearner)
+    def compute_loss_for_module(
+        self,
+        *,
+        module_id: ModuleID,
+        hps: LearnerHyperparameters,
+        batch: NestedDict,
+        fwd_out: Mapping[str, TensorType]
+    ) -> TensorType:
+        if self.module[module_id].is_stateful():
```
Comment (sven1977) [inline rllib/algorithms/bc/torch/bc_torch_learner.py:33]: Can we add a TODO here to fix this zero-padding requirement. What I have in mind is this:

If an algorithm supports RNNs (sequence-aware train batches), we should unify our format requirement and handling of these:
* The train batch should always have the format: (B, T, ...), so it has B rows, each of length T.
* If an episode ends in the middle of a row, we will simply insert a "initial state" at that position instead of doing zero-padding.
* A new field in the batch dict: "is_first" of shape (B, T) will indicate, at which exact B/T-positions to insert these initial states.
* Thus, we will NOT have to do any special reduction anymore. A simple torch.mean should suffice.

We don't have to fix this right now, but I do want to move to this pattern (which DreamerV3 already uses) in the mid-run. So far, this'll affect PPO and BC only.

File: rllib/algorithms/bc/tf/bc_tf_learner.py (hunk 6 above)
Comment (sven1977) [reply rllib/algorithms/bc/tf/bc_tf_learner.py:35]: Also, please add the same TODO here, which I suggested in the torch equivalent code below.

Comment (sven1977) [reply rllib/algorithms/bc/tf/bc_tf_learner.py:35]: On Artur's comment:
This might be difficult as we will have to create an offline dataset for the recurrent case first (e.g. record a PPO + LSTM run vs `StatelessCartPole`).
Maybe we can do this in another PR, but yes, this would be super useful to have.

File: rllib/algorithms/bc/bc.py [hunk 15]
```diff
@@ -73,3 +136,57 @@ class BC(MARWIL):
...
+    def training_step(self) -> ResultDict:
+        if not self.config["_enable_rl_module_api"]:
+            # Using ModelV2.
+            return super().training_step()
+        else:
+            # Implement logic using RLModule and Learner API.
+            with self._timers[SAMPLE_TIMER]:
+                # Sampling from offline data.
+                if self.config.count_steps_by == "agent_steps":
+                    train_batch = synchronous_parallel_sample(
```
Comment (sven1977) [inline rllib/algorithms/bc/bc.py:150]: Add a TODO:
* For offline Algos, we should not use RolloutWorkers (or EnvRunners) anymore.
* Instead, we should handle the (already distributed) dataset directly from within the Algorithm's `training_step`. EnvRunner/RolloutWorker should ONLY be used for actual env stepping and model inference (action computations).

File: rllib/algorithms/bc/tf/bc_tf_rl_module.py (hunk 12 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/tf/bc_tf_rl_module.py:6]: Makes sense to me as the class is simply for tf logic not necessarily for framework setting. 

Will change it also for Torch.

Comment (simonsays1980) [reply rllib/algorithms/bc/tf/bc_tf_rl_module.py:6]: Short note: The `framework = "torch"` etc. was already in the `TorchRLModule`, so nothing to do but remvoing it from the BC Modules.

File: rllib/algorithms/bc/tests/test_bc.py [hunk 16]
```diff
@@ -14,7 +14,7 @@
 class TestBC(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
-        ray.init()
+        ray.init(local_mode=True)
```
Comment (sven1977) [inline rllib/algorithms/bc/tests/test_bc.py:17]:
```suggestion
        ray.init()
```

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 17]
```diff
@@ -75,7 +75,7 @@ class TorchRLModule(nn.Module, RLModule):
     Compiling these methods can bring speedups under certain conditions.
     """
 
-    framwork: str = "torch"
+    framework: str = "torch"
```
Comment (sven1977) [inline rllib/core/rl_module/torch/torch_rl_module.py:78]: Nice catch!

**Review by sven1977 — APPROVED** [review]
LGTM now. Thanks @simonsays1980 !

**Comment by sven1977** [comment]
Pending Tests, then we can merge.

File: rllib/algorithms/bc/bc.py [hunk 18]
```diff
@@ -52,11 +66,60 @@ def __init__(self, algo_class=None):
         # Advantages (calculated during postprocessing)
         # not important for behavioral cloning.
         self.postprocess_inputs = False
+        # Important to be False for dummy runs.
+        self.use_gae = False
```
Comment (sven1977) [inline rllib/algorithms/bc/bc.py:70]: Hey @simonsays1980 , one last question, before we can merge: It seems like use_gae is not used at all anywhere, neither by MARWIL, nor by BC. Should we not completely remove this setting then? Or am I missing something?

Comment (sven1977) [reply rllib/algorithms/bc/bc.py:70]: See this comment here in the MARWIL code:
```
            # We just want the discounted cumulative rewards, so we won't need
            # GAE nor critic (use_critic=True: Subtract vf-estimates from returns).
            use_gae=False,
```
https://github.com/ray-project/ray/blob/master/rllib/algorithms/marwil/marwil_tf_policy.py#L68

**Comment by sven1977** [comment]
One more question before we can merge. See comment on `config.use_gae` for `bc.py`

File: rllib/algorithms/bc/tf/bc_tf_learner.py (hunk 6 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/tf/bc_tf_learner.py:44]: Alright. Makes sense to have a dataset that shows recurrence to apply recurrence successully. Let's move it to another issue. probably the one @ArturNiederfahrenhorst  has opened (#33640).

File: rllib/algorithms/bc/bc.py (hunk 18 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/bc.py:70]: Great catch. I stumbled over it when studyxing the config and thought that the user should have the option to switch over to GAE if needed. B ut you are right, it's never used again. I will erase it.

File: rllib/algorithms/bc/bc.py (hunk 15 above)
Comment (simonsays1980) [reply rllib/algorithms/bc/bc.py:155]: @sven1977 I have opened an issue for this #39849
